bpo-33872: Add list elements access complexity

andresdelfino opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: bpo-33872 -->
https://bugs.python.org/issue33872
<!-- /issue-number -->

andresdelfino committed:
Add list elements access complexity

pablogsal commented on Doc/library/stdtypes.rst:
```diff
@@ -1129,7 +1129,7 @@ Lists
 Lists are mutable sequences, typically used to store collections of
 homogeneous items (where the precise degree of similarity will vary by
-application).
+application).  List elements are accessed with a O(1).
```
Suggestion: "List elements are accessed with O(1) time complexity" or "Accessing list elements has O(1) time complexity".

"List elements are accessed with a O(1)" sounds a bit off to me.

andresdelfino replied:
IMHO, mentioning "time complexity" is definitely an improvement, I'll rephrase, thanks :)

andresdelfino committed:
Update stdtypes.rst

andresdelfino committed:
Update stdtypes.rst

We are not talking much about time complexity in other parts of the documentation. Change it to mimic the intention in the Glossary: avoid the confusion with linked lists.

vadmium commented on Doc/library/stdtypes.rst:
```diff
@@ -1131,6 +1131,9 @@ Lists are mutable sequences, typically used to store collections of
 application).
 
+Despite its name, lists are more akin to arrays in other languages than to
+linked lists, since its elements are accessed with time complexity O(1).
```
Use plural: “_their_ name”, “_their_ elements”, to agree with “lists are”

andresdelfino replied:
:sweat_smile: thanks!

andresdelfino committed:
Address Martin's comment

andresdelfino committed:
Use plural

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -1131,6 +1131,9 @@ Lists are mutable sequences, typically used to store collections of
 homogeneous items (where the precise degree of similarity will vary by
 application).
 
+Despite their name, lists are more akin to arrays in other languages than to
```
“despite their name”  seems redundant.
“List” is datastructure independent, abstract word.
For example, C# uses IList and ArrayList for sequence (indexable). And LinkedList for linked list.
